packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -487,26 +490,34 @@ class IndexedDbRemoteDocumentChangeBuffer extends RemoteDocumentChangeBuffer {
   protected applyChanges(
     transaction: PersistenceTransaction
   ): PersistencePromise<void> {
-    const changes = this.assertChanges();
     let delta = 0;
-    const toApply: Array<{ doc: DbRemoteDocument; key: DocumentKey }> = [];
-    changes.forEach((key, maybeDocument) => {
-      const doc = this.documentCache.serializer.toDbRemoteDocument(
-        maybeDocument
-      );
+    const toApply: Array<{
+      key: DocumentKey;
+      doc?: DbRemoteDocument;
+      size?: number;
+    }> = [];
+
+    this.assertChanges();
+    this.changes!.forEach((key, maybeDocument) => {
       const previousSize = this.documentSizes.get(key);
-      // NOTE: if we ever decide we need to support doing writes without
-      // reading first, this assert will need to change to do the read automatically.
       assert(
         previousSize !== undefined,
         `Attempting to change document ${key.toString()} without having read it first`
       );
-      const size = dbDocumentSize(doc);
-      delta += size - previousSize!;
-      toApply.push({ key, doc });
+      if (maybeDocument) {
+        const doc = this.documentCache.serializer.toDbRemoteDocument(
+          maybeDocument
+        );
+        const size = dbDocumentSize(doc);
+        delta += size - previousSize!;
+        toApply.push({ key, doc, size });
+      } else {
+        delta -= previousSize!;
+        toApply.push({ key });
+      }
     });
 
-    return this.documentCache.addEntries(transaction, toApply, delta);
+    return this.documentCache.modifyEntries(transaction, toApply, delta);
```

Comment:
I'll play with this more but the main reason was that we have to apply the size difference twice. I'll see if I can get a clean version of this, especially since the delta for the deletes will likely be zero.